src/workerd/api/sql-test.js

```diff
@@ -99,28 +99,76 @@ async function test(storage) {
   )
 
   // Test partial query ingestion
-  assert.deepEqual(
-    sql.ingest(`SELECT 123; SELECT 456; -- trailing comment`),
-    23
+  await storage.transaction(async () => {
+    assert.deepEqual(
+      sql.ingest(`SELECT 123; SELECT 456; -- trailing comment`),
+      23
+    )
+    assert.deepEqual(sql.ingest(`SELECT 123; SELECT 456;    `), 23)
+    assert.deepEqual(sql.ingest(`SELECT 123; SELECT 456;`), 23)
+    assert.deepEqual(sql.ingest(`SELECT 123; SELECT 456`), 11)
+    assert.deepEqual(sql.ingest(`SELECT 123; SELECT 45`), 11)
+    assert.deepEqual(sql.ingest(`SELECT 123; SELECT 4`), 11)
+    assert.deepEqual(sql.ingest(`SELECT 123; SELECT `), 11)
+    assert.deepEqual(sql.ingest(`SELECT 123; SELECT`), 11)
+    assert.deepEqual(sql.ingest(`SELECT 123; SELEC`), 11)
+    assert.deepEqual(sql.ingest(`SELECT 123; SELE`), 11)
+    assert.deepEqual(sql.ingest(`SELECT 123; SEL`), 11)
+    assert.deepEqual(sql.ingest(`SELECT 123; SE`), 11)
+    assert.deepEqual(sql.ingest(`SELECT 123; S`), 11)
+    assert.deepEqual(sql.ingest(`SELECT 123; `), 11)
+    assert.deepEqual(sql.ingest(`SELECT 123;`), 11)
+    assert.deepEqual(sql.ingest(`SELECT 123`), 0)
+    assert.deepEqual(sql.ingest(`SELECT 12`), 0)
+    assert.deepEqual(sql.ingest(`SELECT 1`), 0)
+  })
+
+  // Test that sql.ingest checks that it's in a transaction
+  assert.throws(
+    () => sql.ingest(`SELECT 1`),
+    /sql\.ingest must be run inside a transaction/
   )
-  assert.deepEqual(sql.ingest(`SELECT 123; SELECT 456;    `), 23)
-  assert.deepEqual(sql.ingest(`SELECT 123; SELECT 456;`), 23)
-  assert.deepEqual(sql.ingest(`SELECT 123; SELECT 456`), 11)
-  assert.deepEqual(sql.ingest(`SELECT 123; SELECT 45`), 11)
-  assert.deepEqual(sql.ingest(`SELECT 123; SELECT 4`), 11)
-  assert.deepEqual(sql.ingest(`SELECT 123; SELECT `), 11)
-  assert.deepEqual(sql.ingest(`SELECT 123; SELECT`), 11)
-  assert.deepEqual(sql.ingest(`SELECT 123; SELEC`), 11)
-  assert.deepEqual(sql.ingest(`SELECT 123; SELE`), 11)
-  assert.deepEqual(sql.ingest(`SELECT 123; SEL`), 11)
-  assert.deepEqual(sql.ingest(`SELECT 123; SE`), 11)
-  assert.deepEqual(sql.ingest(`SELECT 123; S`), 11)
-  assert.deepEqual(sql.ingest(`SELECT 123; `), 11)
-  assert.deepEqual(sql.ingest(`SELECT 123;`), 11)
-  assert.deepEqual(sql.ingest(`SELECT 123`), 0)
-  assert.deepEqual(sql.ingest(`SELECT 12`), 0)
-  assert.deepEqual(sql.ingest(`SELECT 1`), 0)
 
+  // Test execution of ingested queries by taking an input of 6 INSERT statements, that all
+  // add 6 rows of data, then splitting that into a bunch of chunks, then ingesting them all
+  {
+    sql.exec(`CREATE TABLE streaming(val TEXT);`)
+    const inserts = ['a', 'b', 'c', 'd', 'e', 'f']
+      .map(
+        (c) =>
+          `INSERT INTO streaming VALUES ('${c}a'),('${c}b'),('${c}c'),('${c}d'),('${c}e'),('${c}f');`
+      )
+      .join('\n')
+
+    // Use a chunk size 2, 4, 8, 16, ... characters
```

Comment:
The gold standard for this kind of test is getting 1 byte at a time (so the minimum here should be 1, not 2).

Do we have to worry about encodings here or is something ensuring that we get complete characters?  For example, if we're using UTF-8, what ensures that we're seeing everything in a multi-byte sequence.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String#utf-16_characters_unicode_code_points_and_grapheme_clusters tells me that JavaScript uses UTF-16.  How do we ensure that we don't split up surrogate pairs (as one might have for an emoji in a string)?

Reply:
It's somewhat more accurate to say JavaScript uses WTF-16, that is, it does care if surrogate pairs are split, it just treats strings as arrays of 16-bit values. It's the font renderer's problem to decide what to do about a split surrogate pair.

Meanwhile in KJ land, we use UTF-8 -- or more accurately, WTF-8. If we see properly paired surrogates we will translate them into a single UTF-8 codepoint but if we see unpaired surrogates we'll just translate each one into a WTF-8 codepoint and carry on.

For the purpose of this specific PR, I think none of this is a concern because the only place we might potentially see the input get split in the middle of a code point is at the end of the buffer, but in every such case, we'll simply report that the statement is incomplete.

But as I commented above, it is a problem that the API returns the offset of the end of the last statement as a UTF-8 byte offset, which won't be useful for JavaScript operating on a UTF-16 string.

Reply:
Have refactored this to no longer use character counts at all.

> The gold standard for this kind of test is getting 1 byte at a time (so the minimum here should be 1, not 2).

I've added a test-case for single byte chunks and it's corrupting emojis. See discussion here: https://github.com/cloudflare/workerd/pull/1783#discussion_r1521252427